internal/services/serviceprincipals/service_principal_password_resource_msgraph.go

```diff
@@ -0,0 +1,160 @@
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/manicminer/hamilton/msgraph"
+
+	"github.com/hashicorp/terraform-provider-azuread/internal/clients"
+	helpers "github.com/hashicorp/terraform-provider-azuread/internal/helpers/msgraph"
+	"github.com/hashicorp/terraform-provider-azuread/internal/services/serviceprincipals/parse"
+	"github.com/hashicorp/terraform-provider-azuread/internal/tf"
+)
+
+func servicePrincipalPasswordResourceCreateMsGraph(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics { //nolint
+	client := meta.(*clients.Client).ServicePrincipals.MsClient
+	objectId := d.Get("service_principal_id").(string)
+
+	credential, err := helpers.PasswordCredentialForResource(d)
+	if err != nil {
+		attr := ""
+		if kerr, ok := err.(helpers.CredentialError); ok {
+			attr = kerr.Attr()
+		}
+		return tf.ErrorDiagPathF(err, attr, "Generating password credentials for service principal with object ID %q", objectId)
+	}
+
+	id := parse.NewCredentialID(objectId, "password", *credential.KeyId)
+
+	tf.LockByName(servicePrincipalResourceName, id.ObjectId)
+	defer tf.UnlockByName(servicePrincipalResourceName, id.ObjectId)
+
+	app, status, err := client.Get(ctx, id.ObjectId)
+	if err != nil {
+		if status == http.StatusNotFound {
+			return tf.ErrorDiagPathF(nil, "service_principal_id", "Service principal with object ID %q was not found", id.ObjectId)
+		}
+		return tf.ErrorDiagPathF(err, "service_principal_id", "Retrieving service principal with object ID %q", id.ObjectId)
+	}
+
+	newCredentials := make([]msgraph.PasswordCredential, 0)
+	if app.PasswordCredentials != nil {
+		for _, cred := range *app.PasswordCredentials {
+			if cred.KeyId != nil && *cred.KeyId == *credential.KeyId {
+				return tf.ImportAsExistsDiag("azuread_service_principal_password", id.String())
+			}
+			newCredentials = append(newCredentials, cred)
+		}
+	}
+
+	newCredentials = append(newCredentials, *credential)
+
+	properties := msgraph.ServicePrincipal{
+		ID:                  &id.ObjectId,
+		PasswordCredentials: &newCredentials,
+	}
+	if _, err := client.Update(ctx, properties); err != nil {
+		return tf.ErrorDiagF(err, "Adding password for service principal with object ID %q", id.ObjectId)
+	}
+
+	d.SetId(id.String())
+
+	return servicePrincipalPasswordResourceReadMsGraph(ctx, d, meta)
+}
+
+func servicePrincipalPasswordResourceReadMsGraph(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics { //nolint
+	client := meta.(*clients.Client).ServicePrincipals.MsClient
+
+	id, err := parse.PasswordID(d.Id())
+	if err != nil {
+		return tf.ErrorDiagPathF(err, "id", "Parsing password credential with ID %q", d.Id())
+	}
+
+	app, status, err := client.Get(ctx, id.ObjectId)
+	if err != nil {
+		if status == http.StatusNotFound {
+			return tf.ErrorDiagPathF(nil, "service_principal_id", "Service principal with object ID %q was not found", id.ObjectId)
```

Comment:
in which case the password won't exist, so shouldn't this be removing it from the state?

Reply:
I thought about this, if for some reason the service_principal_id either gets mismatched or there's a 404 for some other reason, the primary copy of the password is in the state, and not auto-removing it could save users from a lot of rotation?